File: lab_11/b.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n;
    std::cin >> n;

    std::vector<long long> ducks(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> ducks[i];
    }

    std::sort(ducks.begin(), ducks.end());

    long long totalCost = 0;
    for (int i = 0; i < n / 2; ++i) {
        totalCost += ducks[i] + ducks[n - i - 1];
        
    }

    for (int i = 0; i < n / 2; ++i) {
        totalCost += ducks[i];
        std::cout << i << std::endl;
    }

    if (n % 2 == 1) {
        totalCost += ducks[0] + ducks[n / 2];
    }

    std::cout << totalCost << std::endl;

    return 0;
}

File: lab_11/b.py
import heapq

def min_cost(n, ducks):
    # Convert list to a heap
    heapq.heapify(ducks)
    total_cost = 0
    while len(ducks) > 1:
        # Pop the smallest element
        a = heapq.heappop(ducks)
        # Add it to the total cost
        total_cost += a
        # Add it to the next smallest element
        b = ducks[0] + a
        # Replace the next smallest element with the new sum
        heapq.heapreplace(ducks, b)
    return total_cost

# Test the function
print(min_cost(5, [2, 5, 4, 2, 8]))  # Output: 27
print(min_cost(5, [5, 5, 5, 5, 5]))  # Output: 40